### PR TITLE
azurerm_*_policy_remediation - should use policy_definition_reference_id in favor of policy_definition_id  Fix

### LINKED ISSUE #16454: Terraform azurerm_resource_group_policy_remediation not working as expected
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

v1.1.7

### AzureRM Provider Version

v3.2.0

### Affected Resource(s)/Data Source(s)

azurerm_resource_group_policy_remediation and remediation tasks

### Terraform Configuration Files

```hcl
data "azurerm_resource_group" "rg_name" {
  name = var.rg_name
}

data "azurerm_subscription" "primary" {
}

resource "azurerm_resource_group_policy_assignment" "policy_assign" {
  name                 = join("_", ["policy_assignment_for", "${var.rg_name}"])
  resource_group_id    = data.azurerm_resource_group.rg_name.id
  policy_definition_id = azurerm_policy_set_definition.inherit_rg_tags.id
  enforce              = true
  identity {
    type = "SystemAssigned"
  }
  location = var.location_for_identity
}

resource "azurerm_role_assignment" "assignment" {
  scope                = join("/", [data.azurerm_subscription.primary.id, "resourceGroups", "${var.rg_name}"])
  #scope                = data.azurerm_subscription.primary.id
  role_definition_name = "Contributor"
  principal_id         = azurerm_resource_group_policy_assignment.policy_assign.identity[0].principal_id
}

output "policy_definition_id_forsnowgrp" {
  value = azurerm_policy_definition.policy["foo_tag"].id
}

resource "azurerm_resource_group_policy_remediation" "example" {
  name                 = "policy_remediation_for_tag_foo_tag_email"
  resource_group_id    = data.azurerm_resource_group.rg_name.id
  policy_assignment_id = azurerm_resource_group_policy_assignment.policy_assign.id
  policy_definition_id = azurerm_policy_definition.policy["foo_tag"].id
}
```

### Debug Output/Panic Output

```shell
Terraform is deeming that it has created the remediation task for "policy_remediation_for_tag_foo_tag_email" But I don't see it in Azure portal. Seems like the terraform module has a bug in remediation task creations.
```

### Expected Behaviour

Terraform should have created the policy_remediation_task for the existing resources of the group.

### Actual Behaviour

The remediation task not gets created in Azure portal

### Steps to Reproduce

Use above provided terraform code snippet and use appropriate variables and run "terraform apply"

### Important Factoids

_No response_

### References

_No response_

### LINKED ISSUE #17096: policy_definition_id in azurerm_management_group_policy_remediation Not Working
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.2.0

### AzureRM Provider Version

3.9.0

### Affected Resource(s)/Data Source(s)

azurerm_management_group_policy_remediation

### Terraform Configuration Files

```hcl
resource "azurerm_management_group_policy_remediation" "main" {
  name                 = "test"
  policy_assignment_id = "/providers/Microsoft.Management/managementGroups/xxx/providers/Microsoft.Authorization/policyAssignments/d70f6f72a0e64d11a63a96b2"
  management_group_id  = "/providers/Microsoft.Management/managementGroups/xxx"
  policy_definition_id = "/providers/Microsoft.Authorization/policyDefinitions/91cb9edd-cd92-4d2f-b2f2-bdd8d065a3d4"
}
```

### Debug Output/Panic Output

```shell
N/A
```

### Expected Behaviour

A new remediation task should be created for this policy definition from the policy set definitions. 

### Actual Behaviour

Terraform apply succeeded, a new TF resource created.
From Azure portal, I was not able to find the remediation task created, also the policy still shows as non-compliant.  
Also "policy_definition_id" argument accepts any string after "policyDefinitions/" which means that it is not actually checking the policy definition id !

### Steps to Reproduce

1. Assign policy set definition to one of Azure Management Groups.
2. Use azurerm_management_group_policy_remediation to create a remediation task for one of the policy definitions under the policy set-definition. 
3. Apply the changes. 
4. Check if a policy remediation task has been created. 

### Important Factoids

_No response_

### References

_No response_

### LINKED ISSUE #17277: azurerm_management_group_policy_remediation fails to remediate resources correctly
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.2.2

### AzureRM Provider Version

3.10.0

### Affected Resource(s)/Data Source(s)

policy remediation

### Terraform Configuration Files

```hcl
resource "azurerm_management_group_policy_remediation" "example1" {
  name                 = "exampleremediation"
  management_group_id  = data.azurerm_management_group.org.id
  policy_assignment_id = azurerm_management_group_policy_assignment.example.id 
}
```

### Debug Output/Panic Output

```shell
It only creates remediation count but does not create remediation task.Hence no trace of remediation
```

### Expected Behaviour

Remediation task should be created in azure portal

### Actual Behaviour

It only appends the remediation count number without creating remediation task

### Steps to Reproduce

resource "azurerm_management_group_policy_remediation" "example1" {
  name                 = "exampleremediation"
  management_group_id  = data.azurerm_management_group.org.id
  policy_assignment_id = azurerm_management_group_policy_assignment.example.id 
}

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by gettek**
Initiative Remediation tasks should use `policy_definition_reference_id` in favor of `policy_definition_id`

Fix #17096
Fix #16454
Fix #17277
Fix https://github.com/gettek/terraform-azurerm-policy-as-code/issues/28

### CODE REVIEW

**Comment by gettek** [comment]
Hey @katbyte any chance of a quick review here please :)

**Comment by wuxu92** [comment]
hi, @gettek thanks for your work and it looks good to me. Just a small tip here, could you please split the updating of the `mode` property's document to a single PR, so this PR only fixes the policy definition id issue?

**Comment by marcelbraak** [comment]
Great job. Cant wait for this PR to be merged.

**Comment by gettek** [comment]
> hi, @gettek thanks for your work and it looks good to me. Just a small tip here, could you please split the updating of the `mode` property's document to a single PR, so this PR only fixes the policy definition id issue?

https://github.com/hashicorp/terraform-provider-azurerm/pull/18181

**Comment by JustinGrote** [comment]
@wuxu92  got time to review this again? :) It's a blocker for me.

**Comment by wuxu92** [comment]
we have to wait for Hashicorp's review, @katbyte could you please have a look at this, thanks.

File: internal/services/policy/remediation_management_group.go [hunk 1]
```diff
@@ -92,12 +92,9 @@ func resourceArmManagementGroupPolicyRemediation() *pluginsdk.Resource {
 				},
 			},
 
-			"policy_definition_id": {
+			"policy_definition_reference_id": {
```
Comment (katbyte) [inline internal/services/policy/remediation_management_group.go:95]: Thanks @gettek, however we cannot just remove the old property/ we'll need to deprecate it and then wrap it in a features.FourPointOhBeta flag.

File: internal/services/policy/remediation_resource.go [hunk 2]
```diff
@@ -91,12 +91,9 @@ func resourceArmResourcePolicyRemediation() *pluginsdk.Resource {
 				},
 			},
 
-			"policy_definition_id": {
+			"policy_definition_reference_id": {
```
Comment (katbyte) [inline internal/services/policy/remediation_resource.go:94]: and here

**Review by katbyte — CHANGES_REQUESTED** [review]
@gettek - can we please keep the old property and deprecate it letting users know to move to the new one. if you search for features.FourPointOhBeta you can see a number of examples on how to do this.
this way no ones configs are broken and it'll automatically be removed when we release 4.0

thanks!

File: internal/services/policy/remediation_management_group.go (hunk 1 above)
Comment (gettek) [reply internal/services/policy/remediation_management_group.go:95]: Thanks @katbyte properties added back with deprecation message

Comment (gettek) [reply internal/services/policy/remediation_management_group.go:95]: Thanks @katbyte properties added back with deprecation message

**Review by gettek — COMMENTED** [review]
Have also removed `policy_definition_id` from tests, which should be `policy_definition_reference_id`. This property is only for **initiative** assignments but these tests only run **definition** assignments

**Comment by corticalstack** [comment]
Hey. Any expected timeline when this fix will be merged into hashicorp:main, and with what version of hashicorp/terraform-provider-azurerm? Thanks

**Review by katbyte — APPROVED** [review]
Thanks @gettek - LGTM now! 🌩️

**Pull request merged**